### ISSUE TITLE
Extend the unleash interface with context-data

### ISSUE THREAD

**Issue by ivarconr** [Unleash/unleash-client-java#29]
In the current version of Unleash strategies needs to get their context data via special provides which typically is implemented as some magic-data available at the current request. This works well in most java-based web-applications. 

However doing toggles in a standalone web-application this becomes much harder as you typically do not have a user bound to the current request. In this scenario you will typically have the context-data available runtime, at the same location you want to verify if a toggle should be on or of. 

It would thus be beneficial if the client also could inject context data at runtime. The unleash API would then be overloaded with a new method, the the user can inject the required context-data. 

``` java
unleash.isEnabled("app.someToggle", UnleashConext.of("k1", "v1", "k2", "v2"));
```

Unleash will then in turn inject the context-data to the strategy which would be extended to:

``` java
public interface Strategy {
    String getName();

    boolean isEnabled(Map<String, String> parameters);

    boolean isEnabled(Map<String, String> parameters, UnleashConext context);
}
```

(\* if we move the client to java8 we could also provide a default implementation of the new method to avoid breaking the API)

This is also the approach used for the [unleash-client-node](https://github.com/finn-no/unleash-client-node).

**Comment by ivarconr** [comment]
(from https://github.com/finn-no/unleash/issues/99)

**Comment by ivarconr** [comment]
After years of usage i finn we see that a UnleashUser with the following properties would make it simple to standardize a lot of strategies (all fields are of course optional):

```
UnleashUser::
 identifier: String,
 remoteAddress: String,
 sessionId: String,
 properties: Map<String, String>
```

The UnleashUser should be part of the `UnleashContext` and it should be possible to provide the context via a request-bound provider (e.g. in a request scoped spring bean, or similar). It should also be possible to send in the `UnleashContext` (with an unleashUser on it) as part of the "isEnabled" call. This is very convenient in non-thread based environment where the user/request details is available just in place.

**Comment by ivarconr** [comment]
Suggestion: 

should we change the `properties` field to be a just a list of flags instead? The only use case we have for it at the moment us to flag a user as a beta-user. This feels a bit cumbersome in a key,value structure. 

Aka:

```UnleashContext::
 userId: String,
 sessionId: String,
 remoteAddress: String,
 flags: Array<String>
```

What do you think @sveisvei  and @vsandvold ?

**Comment by vsandvold** [comment]
Key value feels a bit awkward, but is more flexible of course. For the abtest strategy I thought about providing context for device, referrer and other useful properties for segmentation. Currently it's only the USERID cookie value that is provided.

Another thing... Could it be useful to have multiple context providers, like BetaUserContextProvider and ABTestContextProvider?

**Comment by ivarconr** [comment]
> Currently it's only the USERID cookie value that is provided

yes, then key/value is required i suppose, better keep the properties then. 

>  Could it be useful to have multiple context providers, like BetaUserContextProvider and ABTestContextProvider?

No, the whole purpose is to provide one context and one entry point for it! It's better to make it flexible enough to add what we need, IMHO.

**Comment by ivarconr** [comment]
solved in #33
